### DISCUSSION TITLE
gNB does not work as shown in Documentation

### DISCUSSION THREAD

**Discussion by oguztudo** [srsran/srsRAN_Project#88] [General Help]
Hello everybody,

I am running the gNB and i dont know if this is an issue or if this works as intended but my gNB sends signals irregularly unlike the screenshot posted here:

https://docs.srsran.com/projects/project/en/latest/_images/maia_sdr.png

Mine looks like this:

![gNB_flow](https://github.com/srsran/srsRAN_Project/assets/129198919/89ea1895-dd09-4cd4-a61d-8fb166d6c087)

Detecting the gNB works on my Samsung Galaxy S23 Ultra in the NR only mode as described in the tutorial. However i can not connect successfully. The first two lines of the log confuse me a little because one says rlc_mode=UM and the other one says AM (I suppose the rlc mode gets changed to AM). 
There are some lines which say _"criticality": "reject"_ but I don't know if that's a Problem causing factor.
I think the connection could be established solving the the Real-time failure in FAPI issue which can be seen further down in the log.

Here is my config and the log to gNB with detail level debug for all:

[cots_gnb.txt](https://github.com/srsran/srsRAN_Project/files/11651906/cots_gnb.txt)

[gnb.log](https://github.com/srsran/srsRAN_Project/files/11652067/gnb.log)

**Comment by suttonpd** [comment]
Your gnb log shows real-time failures for every DL TTI - ~70 real-time failures in <2ms. 
What are the specs of the machine you're trying to use?

**Reply by oguztudo** [reply]
Hello Paul,

I am using a USRP X310 with UHD 4.0.0 running a XG image flavor over a SFP+ 10Gbit Port.

My PC runs on Kubuntu 20.04 LTS running on a Intel XEON W-2245

My UE is a Samsung Galaxy S23 Ultra running on a Qualcomm Snapdragon 8 Gen 2

**Reply by ismagom** [reply]
Can you try removing the num_recv and num_send_frames from the uhd args? Also remove otw_format and time_calibration. Those are all configs for the B210.

**Reply by ismagom** [reply]
Couple more things: make sure you pull the latest code before trying again, we update it every 2 weeks. And run the srsran_performance script under scripts/ folder. 

Finally, when you attach logs, set log level to info.

**Reply by oguztudo** [reply]
Hello @ismagom ,
Thank you for the reply. So i tried both of your suggestions and will split my answer in 2 Parts.

**Part 1:**
I removed those from the UHD args but the issue persists. How can I know which configs are device exclusives?
I also try to have the settings of the performance script active when i work with srsRAN and the USRP.

here is the output of the gNB:
![gnb_usrp](https://github.com/srsran/srsRAN_Project/assets/129198919/1a82e570-39dd-4093-b2d6-f08449e9679a)

The log of the gNB on info level:
![gnb_log](https://github.com/srsran/srsRAN_Project/assets/129198919/cd02b2e8-1228-42b8-901e-ddb2941fa30a)

And the Tail of the AMF status:
![amf_tail](https://github.com/srsran/srsRAN_Project/assets/129198919/44121b32-1035-4cdc-8b2f-c1bdbb4f69a1)


**Part 2:**
After i pulled the latest code i get the following
 lines in the console when i try to start the gNB:
![latest](https://github.com/srsran/srsRAN_Project/assets/129198919/ac1323e5-9065-46c4-acea-ab30f46a107a)
Downloading srsRAN Project 23.5 from Github and installing it works but i suppose that is not the newest version. Is this Problem related to the OFH support that has been added?

**Reply by brendan-mcauliffe** [reply]
With the 23.5 release, we have updated the configuration file. Instead of `rf_driver`, change the section name to `ru_sdr`. We will provide updated documentation for the configuration file changes soon. 

You can also run `sudo ./gnb --help` to get a full list of the accepted configuration file parameters and section names.

**Reply by oguztudo** [reply]
Hello @brendan-mcauliffe,
thanks for the info, that did do the trick to make it work BUT i still encounter the issue of not being able to connect.
I realized that the log became more detailed on the info setting so i will post the log file here:

[gnb.log](https://github.com/srsran/srsRAN_Project/files/11664948/gnb.log)

**Reply by brendan-mcauliffe** [reply]
This log only seems to be running for a few seconds. I recommend you try connecting the phone and sending the logs from that test. 

Are you using an external source for clocking the gNB?

**Reply by oguztudo** [reply]
I do not have a GPSDO but I do have a gps antenna on the GPS ANT Port and a 10MHz Signal from a Signal generator in the REF IN Port of my USRP x310.
Despite not having a GPSDO I still tried all possible variations on the parameters (internal, external, gpsdo) clock and sync but none of those provided any change in the outcome.

This time i tried to run it for a longer period of time and i am trying to give as much detail as possible:
with clock and sync set to internal.
I ran the gNB for 10 mins now. After some tries i could find the gNB, which was almost at min from start of the gNB. Then i tried to connect to the gNB. The whole process of trying to register until the connection fail last about 2 and a half minutes. Then i made another try to register to the gNB with my phone and again nothing. Then i made another search, found the gNB again. Trying to connect happened at minute 10 since starting the gNB, lasted another 2min 15s and ended in a connection fail yet again. This is the log, the amount of information is the same.

[gnb.log](https://github.com/srsran/srsRAN_Project/files/11665817/gnb.log)

I made another run with clock and sync on external but the outcome is the same...

**Reply by ismagom** [reply]
Can you try commenting out the parameters `time_alignment_calibration: -300` and `otw_format: sc12`. 

Also try reducing the `tx_gain:` to `20`. 

Which is IMSI you have configured in the SIM card? 

I think most likely there is a clocking issue. Using the external REF IN is not enough without a PPS (I think), so I'm not sure you can use `external` parameter. 

You say you don't have a GPSDO, but I see in the console the message: `Found an internal GPSO: LC_XO` so in theory setting `gpsdo` to both sync and clock should pick it up. 

Do you have some external instrument to measure the LO offset? 

Regards

**Reply by oguztudo** [reply]
I deleted those parameters already in my earlier tries when Brendan suggested me to do it.
Here is the latest config i used:

[cots_gnb.txt](https://github.com/srsran/srsRAN_Project/files/11676888/cots_gnb.txt)

Changing the tx gain did not provide any solution either (i also varied the distance of the UE to the gNB).

I used he following IMSI, starting with a 00101 and then used something random: 001010000082271

Regarding the GPSDO: Yes the USRP has an internal GPSDO but the precision of 2*10^⁻8 seemed low to me when compared to the Leo Bodnar GPSDO, which is being advertised with running at a precision of 1*10^-12. Also i can only set the clock to GPSDO, when i try to set the sync it says "Error: failed to read sensors" which appears to be a hardware problem (?). So for clock only internal and external are possible.

My spectrum analyzer gives me a maximum for the local oscillator at the set dl_freq. I also could get my hands on a fieldfox today which can show me 5G NR cells. There i could see the SSB offset being mostly at -60kHz and sometimes at 0kHz. The fieldfox could only see the cell properly after setting the frequency offset to +60kHz. In this case i could see the whole spectrum shift by 60kHz (LO included) and the fieldfox could find the cell instantly. I could even detect a change in pci, but eventually set it back to 1. Then i tested and tried to see if i could find the gNB with my phone but could not find anything...

**Comment by oguztudo** [comment]
Update:
I went to a non Roaming scenario, setting the gNB PLMN to 00101 and also turned Roaming and VoLTE off on my phone. I could manage to establish a connection!
Nonetheless i can not access the internet via my phone, despite having the settings from the open5gs site active... I guess thats the next step.

**Reply by brendan-mcauliffe** [reply]
If you cannot access the internet, it is probably a misconfiguration with the APN. Double-check how the APN is configured on your phone and in the core. Make sure they match exactly.

**Reply by oguztudo** [reply]
My phone APN is exactly the same as the one given in the docu. The core APN has the following settings. When i set a UE IPv4 Address, i can ping that corresponding address successfully so maybe i am missing something here...

heres my APN setting in the core:
![image](https://github.com/srsran/srsRAN_Project/assets/129198919/6f48ff93-ccd7-468d-9dec-5513efc7faa1)
![image](https://github.com/srsran/srsRAN_Project/assets/129198919/25f48118-dd2c-440b-91b4-2608733caed8)

**Reply by brendan-mcauliffe** [reply]
Check the consistency between the phone and APN again, once you are sure they're the same try restarting the open5GS services and try to connect again.

**Reply by oguztudo** [reply]
I set the UE and SMF IPAdresses and managed to make it work! Thank you very much for the help!!!
